### PR TITLE
FYST 519 Create role and permissions for NJ in The Hub

### PR DESCRIPTION
**Pull request by mrotondo**
## Link to pivotal/JIRA issue
- https://codeforamerica.atlassian.net/browse/FYST-519
## Is PM acceptance required? (delete one)
- Yes - don't merge until JIRA issue is accepted!
## What was done?
- Adds a "FYST Admin Tools" section to the nav bar of the hub, splitting out the previous "State File Actions" entries in the Admin Tools 
- Creates a new `StateFileNjStaffRole` role, and grant it permission to manage NJ EfileErrors, EfileSubmissions and FaqCategories/FaqItems
- Restrict non-NJ staff from managing NJ EfileErrors, EfileSubmissions or FaqCategories/FaqItems (this last one I'm unsure about, as it wasn't explicitly in the story requirements, so I may change it back)
- Changes some queries to either start with cancancan's loaded resource, or use `accessible_by` if they are "manual" queries that don't start with cancancan's loaded resource
- Use `authorize!` in a few more places to ensure that the currently signed-in user can take certain actions
- Change FaqCategories controller to load and show all our active states automatically, instead of listing them out manually
- Added an NJ FAQ Category product type

Assorted other:
- reduces our usage of explicit before actions like :require_state_file and :require_admin in favor of more specific usages of cancancan's abilities
- fixes some bugs with the FAQ categories and items forms 
  - error messages wouldn't show up in the flash bar at the top
  - on submission of `/new` with validation errors, the product type would disappear and the form would become unusable
  - on submission of `/edit` with validation errors, the product type would switch from the correct state to gyr
- changed a few constraints in the ability.rb file to hopefully be more performant
## How to test?
1. Go to the hub on this heroku deployment
2. Sign in as an admin
3. Invite another email you control as a New Jersey Staff role
4. Accept that invite, go to the hub and sign in, then click FYST Admin Tools and click around to see if things work like they're supposed to
## Screenshots (for visual changes)
image
image
image

### CODE REVIEW

**Review by powersurge360 — APPROVED** [review]
Looks good. I have complaints but they don't originate with this PR and are out of scope. Didn't test but everything seems to make sense.

File: app/models/user.rb [hunk 1]
```diff
@@ -237,6 +238,10 @@ def coalition_lead?
     role_type == CoalitionLeadRole::TYPE
   end
 
+  def state_file_nj_staff?
+    role_type == StateFileNjStaffRole::TYPE
+  end
```
[the remark this answers is not in the document]
Comment (mrotondo) [reply app/models/user.rb:243]: Agreed! Want to write a one sentence description of your idea [here](https://www.notion.so/cfa/106373fd79b280428e9ae0c556f7e4d7?v=a3c648bb70274f54a817118dab6e76ef), or add it to [my entry](https://www.notion.so/cfa/106373fd79b280428e9ae0c556f7e4d7?v=a3c648bb70274f54a817118dab6e76ef&p=ae623a8cb3ed4cb9bfd4c63ee1471265&pm=s) if you think they overlap enough?

File: app/lib/ability.rb [hunk 2]
```diff
@@ -22,16 +22,24 @@ def initialize(user)
...
           cannot :manage, intake_class
         end
         # Enumerate classes here...
+        cannot :manage, StateFile::AutomatedMessage
         cannot :manage, StateFile1099G
         cannot :manage, StateFileDependent
         cannot :manage, StateFileW2
         cannot :manage, StateId
-        cannot :manage, EfileSubmission, id: EfileSubmission.for_state_filing.pluck(:id)
+        cannot :manage, EfileSubmission, data_source_type: StateFile::StateInformationService.state_intake_class_names
```
[the remark this answers is not in the document]
Comment (mrotondo) [reply app/lib/ability.rb:42]: Agains, strongly agree

File: app/controllers/hub/state_file/efile_errors_controller.rb [hunk 3]
```diff
@@ -39,8 +39,8 @@ def update
     def reprocess
       if @efile_error.present? && (@efile_error.auto_wait || @efile_error.auto_cancel)
         auto_transition_to_state = @efile_error.auto_wait ? :waiting : :cancelled
-        submission_ids = EfileSubmissionTransitionError.includes(:efile_error, :efile_submission_transition).where(efile_error: @efile_error, efile_submission_transitions: { most_recent: true, to_state: ["rejected", "failed"] }).pluck(:efile_submission_id)
-        EfileSubmission.where(id: submission_ids).find_each { |submission| submission.transition_to(auto_transition_to_state) }
+        submission_ids = EfileSubmissionTransitionError.accessible_by(current_ability).includes(:efile_error, :efile_submission_transition).where(efile_error: @efile_error, efile_submission_transitions: { most_recent: true, to_state: ["rejected", "failed"] }).pluck(:efile_submission_id)
+        EfileSubmission.accessible_by(current_ability).where(id: submission_ids).find_each { |submission| submission.transition_to(auto_transition_to_state) }
```
[the remark this answers is not in the document]
Comment (mrotondo) [reply app/controllers/hub/state_file/efile_errors_controller.rb:43]: Yes! Very good for batch processes that you don't want to run out of memory (or take forever before processing anything) due to the performance of one giant query.

File: spec/controllers/hub/state_file/efile_submissions_controller_spec.rb [hunk 4]
```diff
@@ -36,6 +44,17 @@
...
       end
     end
+
+    context "with an authenticated non-state file admin" do
+      before { sign_in(create(:admin_user)) }
+
+      it "returns 403 forbidden" do
+        get :show, params: params
+
+        expect(response.status).to eq 403
```
[the remark this answers is not in the document]
Comment (mrotondo) [reply spec/controllers/hub/state_file/efile_submissions_controller_spec.rb:54]: I like that - I'll plop it in here.

**Pull request merged**